core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/DefaultRoutesLoader.java

```diff
@@ -0,0 +1,103 @@
+ * limitations under the License.
+ */
+package org.apache.camel.impl.engine;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.CamelContextAware;
+import org.apache.camel.ExtendedCamelContext;
+import org.apache.camel.RoutesBuilder;
+import org.apache.camel.spi.FactoryFinder;
+import org.apache.camel.spi.Resource;
+import org.apache.camel.spi.RoutesBuilderLoader;
+import org.apache.camel.spi.RoutesLoader;
+import org.apache.camel.spi.annotations.JdkService;
+import org.apache.camel.util.FileUtil;
+import org.apache.camel.util.ObjectHelper;
+
+@JdkService(RoutesLoader.FACTORY)
+public class DefaultRoutesLoader implements RoutesLoader {
+    private CamelContext camelContext;
+
+    public DefaultRoutesLoader() {
+    }
+
+    public DefaultRoutesLoader(CamelContext camelContext) {
+        this.camelContext = camelContext;
+    }
+
+    @Override
+    public CamelContext getCamelContext() {
+        return camelContext;
+    }
+
+    @Override
+    public void setCamelContext(CamelContext camelContext) {
+        this.camelContext = camelContext;
+    }
+
+    @Override
+    public Collection<RoutesBuilder> findRoutesBuilders(Collection<Resource> resources) throws Exception {
+        List<RoutesBuilder> answer = new ArrayList<>(resources.size());
+
+        for (Resource resource : resources) {
+            // language is derived from the file extension
+            final String extension = FileUtil.onlyExt(resource.getLocation(), true);
+
+            if (ObjectHelper.isEmpty(extension)) {
+                throw new IllegalArgumentException("Unable to determine extension for resource: " + resource.getLocation());
+            }
+
+            answer.add(getLoader(extension).loadRoutesBuilder(resource));
+        }
+
+        return answer;
+    }
+
+    /**
+     *
+     * @param  extension
+     * @return
+     */
+    private RoutesBuilderLoader getLoader(String extension) {
+        RoutesBuilderLoader answer = getCamelContext().getRegistry().lookupByNameAndType(extension, RoutesBuilderLoader.class);
+
+        if (answer == null) {
+            final ExtendedCamelContext ecc = getCamelContext(ExtendedCamelContext.class);
+            final FactoryFinder finder = ecc.getFactoryFinder(RoutesBuilderLoader.FACTORY_PATH);
+
+            final BaseServiceResolver<RoutesBuilderLoader> resolver
+                    = new BaseServiceResolver<>(extension, RoutesBuilderLoader.class, finder);
+            final Optional<RoutesBuilderLoader> loader
+                    = resolver.resolve(ecc);
+
+            if (loader.isPresent()) {
+                return CamelContextAware.trySetCamelContext(loader.get(), ecc);
+            }
+        }
+
+        if (answer == null) {
+            throw new IllegalArgumentException("Unable to fina a RoutesBuilderLoader for extension " + extension);
```

Comment:
Typo

Reply:
Also extension may be ambigious, is it quarkus extension or what? Maybe say file extension